ouroboros-network/src/Ouroboros/Network/Protocol/VersionNegotiation/Test.hs

```diff
@@ -0,0 +1,386 @@
+
+module Ouroboros.Network.Protocol.VersionNegotiation.Test where
+
+import           Control.Monad.ST (runST)
+import           Data.Typeable (Typeable, cast)
+import           Data.List (intersectBy, maximumBy, nub, nubBy, sortBy)
+import           Data.Function (on)
+import           Data.Maybe (isJust)
+import           GHC.Generics
+
+import           Control.Monad.IOSim (runSimOrThrow)
+
+import           Network.TypedProtocol.Codec
+import           Network.TypedProtocol.Proofs
+
+import           Ouroboros.Network.Protocol.VersionNegotiation.Type
+import           Ouroboros.Network.Protocol.VersionNegotiation.Codec
+
+import qualified Codec.CBOR.Encoding as CBOR
+import qualified Codec.CBOR.Decoding as CBOR
+import           Codec.Serialise (Serialise)
+import qualified Codec.Serialise     as CBOR
+
+import           Test.QuickCheck
+import           Test.Tasty (TestTree, testGroup)
+import           Test.Tasty.QuickCheck (testProperty)
+
+tests :: TestTree
+tests =
+  testGroup "Ouroboros.Network.Protocol.VersionNegotiation"
+  [ testProperty "connect"             prop_connect
+  , testProperty "codec"               prop_codec_VersionNegotiation
+  , testGroup "Generators"
+    [ testProperty "ArbitraryVersions" $
+        checkCoverage prop_arbitrary_ArbitraryVersions
+    , testProperty "arbitrary ArbitraryValidProtocolVersions"
+        prop_arbitrary_ArbitraryValidProtocolVersions
+    , testProperty "shrink ArbitraryValidProtocolVersions"
+        prop_shrink_ArbitraryValidProtocolVersions
+    ]
+  ]
+
+--
+-- Test Versions
+--
+-- Notes: Associated data are choosen in such a way that a decoder will fail
+-- interpreting one of them as the other.  This is done on purpose for testing
+-- missencoded data (protocol version & associated version data mismatch)
+--
+
+-- |
+-- Testing version number
+--
+data VersionNumber
+  = Version_0
+  | Version_1
+  | Version_2
+  deriving (Eq, Ord, Enum, Bounded, Show)
+
+instance Arbitrary VersionNumber where
+  arbitrary = oneof (map return [minBound .. maxBound])
+
+instance Serialise VersionNumber where
+  encode Version_0 = CBOR.encodeWord 0
+  encode Version_1 = CBOR.encodeWord 1
+  encode Version_2 = CBOR.encodeWord 2
+  decode = do
+    x <- CBOR.decodeWord
+    case x of
+      0 -> return Version_0
+      1 -> return Version_1
+      2 -> return Version_2
+      _ -> fail "decode VersionNumber: wrong tag"
+
+
+-- |
+-- Data Associated with @'Version_0'@
+--
+data Data_0 = C0 | C1 | C2
+  deriving (Eq, Show, Typeable, Generic)
+
+instance Serialise Data_0 where
+  encode C0 = CBOR.encodeWord 0
+  encode C1 = CBOR.encodeWord 1
+  encode C2 = CBOR.encodeWord 2
+  decode = do
+    n <- CBOR.decodeWord
+    case n of
+      0 -> return C0
+      1 -> return C1
+      2 -> return C2
+      _ -> fail "decode Delta_0: wrong tag"
+
+instance Arbitrary Data_0 where
+  arbitrary = oneof [return C0, return C1, return C2]
+  shrink C2 = [C1, C0]
+  shrink C1 = [C0]
+  shrink C0 = []
```

Comment:
Is this really what we want for shrink? I would have thought these versions were "flat", in that neither one is less complex than any other.

Reply:
Oh never mind this is data for one version. Still though, each constructor seems to be of the same complexity.

Reply:
An arbitrary shink order is fine for things like this imho.